README.md
Webmake middleware for express and connect
==========================================

This module is a simple helper to make development with [webmake][1] straight-forward to set up. Webmake is a lightweight client-side dependency management tool that allows you to use node modules in the browser. You can read up on it [here][1].

This module is a simple middleware for [connect][2] and its cousin [express][3].

## To install

    npm install webmake-middleware

## Usage

If you are already using `express` or `connect` for your app, simply configure your server to use the middleware. Since this middleware is currently only for use during development, make sure you put it into the `development` configuration of your `express` application! In the future, there is a good chance that it can be configured to be used for production as well so that it doesn't re-compile the scripts on each request and that the cache headers are configurable.

```javascript
var connect = require('connect'); // this could also be express
var webmakeMiddleware = require('webmake-middleware');

// the `/build.js` url is the request url. `/path/to/main.js` is the path to the
// main entry point.
var server = connect().use(webmakeMiddleware({
  '/build.js': '/path/to/main.js'
  '/build2.js': '/path/to/main2.js'
}));
// add other middleware as necessary

server.listen(3000);
```

## Tests

    # in the webmake-middleware directory
    npm install
    npm test



[1]: http://github.com/medikoo/modules-webmake
[2]: http://www.senchalabs.org/connect/
[3]: http://expressjs.com


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.0.0"></a>
# [1.0.0](https://github.com/gillesruppert/webmake-middleware/compare/v0.1.0...v1.0.0) (2018-11-09)


### Chores

* bump dependencies ([06e385e](https://github.com/gillesruppert/webmake-middleware/commit/06e385e))


### Code Refactoring

* update up to new lint rules ([c309075](https://github.com/gillesruppert/webmake-middleware/commit/c309075))


### BREAKING CHANGES

* Drop support for Node.js version below v4
* Drop support for node.js verions below v0.10


test/helpers.js
"use strict";

const expect = require("expect.js");

exports.successResponseCallback = function (cb, close) {
	return function (res) {
		res.setEncoding("utf8");
		expect(res.statusCode).to.be(200);
		expect(res.headers["content-type"]).to.be("text/javascript; charset=utf-8");
		expect(res.headers["cache-control"]).to.be("no-cache");

		res.on("data", chunk => {
			expect(chunk).to.be.a("string");
			expect(chunk).to.not.be.empty();
		});

		res.on("end", () => {
			if (close) close();
			cb();
		});
	};
};

exports.failureResponseCallback = function (cb, close) {
	return function (res) {
		res.setEncoding("utf8");
		expect(res.statusCode).to.be(200);

		res.on("data", chunk => {
			expect(chunk).to.be.a("string");
			expect(chunk).to.contain("Cannot require");
		});

		res.on("end", () => {
			if (close) close();
			cb();
		});
	};
};


test/connect-test.js
"use strict";

const connect = require("connect");
const http = require("http");
const { normalize } = require("path");
const async = require("async");

const webmakeMw = require("../");
const helpers = require("./helpers");

describe("connect middleware", () => {
	it("should return the compiled file when requesting a valid url ", done => {
		const path = normalize(`${ __dirname }/fixtures/main.js`)
		    , server = connect().use(webmakeMw({ "/test.js": path })).listen(51234, err => {
				if (err) throw err;
				http.get(
					"http://localhost:51234/test.js",
					helpers.successResponseCallback(done, server.close.bind(server))
				);
			});
	});

	it(
		"should return a 500 response when webmake has a problem with compiling the script",
		done => {
			const path = normalize(`${ __dirname }/fixtures/err.js`)
			    , server = connect().use(webmakeMw({ "/error.js": path })).listen(51234, err => {
					if (err) throw err;
					http.get(
						"http://localhost:51234/error.js",
						helpers.failureResponseCallback(done, server.close.bind(server))
					);
				});
		}
	);

	it("should be configurable with an object", done => {
		const config = {
				"/test1.js": normalize(`${ __dirname }/fixtures/main.js`),
				"/test2.js": normalize(`${ __dirname }/fixtures/page2.js`)
			}
		    , server = connect().use(webmakeMw(config)).listen(51234);

		async.parallel(
			[
				function (cb) {
					http.get(
						"http://localhost:51234/test1.js", helpers.successResponseCallback(cb)
					);
				},
				function (cb) {
					http.get(
						"http://localhost:51234/test2.js", helpers.successResponseCallback(cb)
					);
				}
			],
			err => {
				if (err) throw err;
				server.close(done);
			}
		);
	});
});


test/express-test.js
"use strict";

const express = require("express");
const http = require("http");
const { normalize } = require("path");
const async = require("async");

const webmakeMw = require("../");
const helpers = require("./helpers");

describe("express framework", () => {
	it("should return the compiled file when requesting a valid url ", done => {
		const path = normalize(`${ __dirname }/fixtures/main.js`)
		    , server = express().use(webmakeMw({ "/test.js": path })).listen(51234, err => {
				if (err) throw err;
				http.get(
					"http://localhost:51234/test.js",
					helpers.successResponseCallback(done, server.close.bind(server))
				);
			});
	});

	it(
		"should return a 500 response when webmake has a problem with compiling the script",
		done => {
			const path = normalize(`${ __dirname }/fixtures/err.js`)
			    , server = express().use(webmakeMw({ "/error.js": path })).listen(51234, err => {
					if (err) throw err;
					http.get(
						"http://localhost:51234/error.js",
						helpers.failureResponseCallback(done, server.close.bind(server))
					);
				});
		}
	);

	it("should be configurable with an object", done => {
		const config = {
				"/test1.js": normalize(`${ __dirname }/fixtures/main.js`),
				"/test2.js": normalize(`${ __dirname }/fixtures/page2.js`)
			}
		    , server = express().use(webmakeMw(config)).listen(51234);

		async.parallel(
			[
				function (cb) {
					http.get(
						"http://localhost:51234/test1.js", helpers.successResponseCallback(cb)
					);
				},
				function (cb) {
					http.get(
						"http://localhost:51234/test2.js", helpers.successResponseCallback(cb)
					);
				}
			],
			err => {
				if (err) throw err;
				server.close(done);
			}
		);
	});
});


index.js
"use strict";

const count     = require("es5-ext/object/count")
    , isValue   = require("es5-ext/object/is-value")
    , validate  = require("es5-ext/object/valid-value")
    , { parse } = require("url")
    , webmake   = require("webmake")
    , encode    = require("entities").encodeXML
    , errorTpl  = require("fs").readFileSync(`${ __dirname }/error-template.js`, "utf8");

const { stringify } = JSON;

module.exports = function (config, options = {}) {
	const log = isValue(options.log) ? options.log : true;
	delete options.log;
	config = Object(validate(config));

	if (!isValue(options.cache)) options.cache = true;
	return function (req, res, next) {
		const path = parse(req.url).pathname;
		if (!isValue(config[path])) {
			next();
			return;
		}

		const promise = webmake(config[path], options, (err, src) => {
			let msg, modulesSize, packagesSize;

			res.writeHead(200, {
				"Content-Type": "text/javascript; charset=utf-8",
				"Cache-Control": "no-cache"
			});

			if (err) {
				// fail loudly on the cli and in the browser
				console.error(`Webmake error: ${ err.message }`);
				res.end(
					errorTpl
						.replace("$PATH", stringify(encode(path)).slice(1, -1))
						.replace("$ERROR", stringify(encode(err.message)).slice(1, -1))
				);
				return;
			}

			res.end(src);
			if (log) {
				modulesSize = promise.parser.modulesFiles.length;
				packagesSize = count(promise.parser.packages);
				msg = `Webmake OK [${ modulesSize } module`;
				if (modulesSize > 1) msg += "s";
				msg += ` from ${ packagesSize } package`;
				if (packagesSize > 1) msg += "s";
				console.log(`${ msg } in ${ (promise.time / 1000).toFixed(2) }s] -> ${ req.url }`);
			}
		});
	};
};


lib/webmake-middleware.js
var webmake = require('webmake');

/**
 * configuration function
 * Takes either:
 * - 2 strings, 1 being the requested url and the other being
 *   the path to the entry file, i.e. main.js
 * or
 * - 1 object where the key is the url and the value is entry file. This is
 *   useful if you have different urls and entry points
 *
 * @param {string|object} url
 * @param {string} [file]
 * @return {function} middleware
 */
module.exports = function(url, file) {
  var l = arguments.length;
  var config;
  // configuration must be either 2 strings or 1 object
  if (!(l === 2 && typeof url === 'string' && typeof file === 'string' ||
        l === 1 && typeof url === 'object') )  throw new TypeError('wrong arguments passed');

  if (l === 2) {
    config = {};
    config[url] = file;
  }
  else if (l === 1) config = url;

  return function webmakeMiddleware(req, res, next) {
    if (! (req.url in config)) {
      next();
      return;
    }

    webmake(config[req.url], { sourceMap: true }, function(err, src) {
      if (err) {
        // fail loudly on the cli and in the browser
        console.error(err);
        res.writeHead(500);
        res.end('document.write(\'<div style="position:absolute;' +
                'top:10px;width:80%;color:red;font-size:150%;background-color:white;' +
                'border:5px solid black;padding:10px;font:sans-serif;">' +
                err.message.replace(/'/g, '\\\'') +
               '</div>\')');
        return;
      }

      res.writeHead(200, {
        'Content-Type': 'text/javascript; charset=utf-8',
        'Cache-Control': 'no-cache'
      });

      res.end(src);
    });
  };
};
